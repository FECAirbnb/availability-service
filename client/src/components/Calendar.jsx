/* eslint-disable no-plusplus */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable react/no-access-state-in-setstate */
/* eslint-disable react/destructuring-assignment */
import React from 'react';
import {
  compareAsc,
  format,
  startOfWeek,
  addDays,
  startOfMonth,
  endOfMonth,
  endOfWeek,
  isSameMonth,
  isSameDay,
  subMonths,
  addMonths
} from 'date-fns';

class Calendar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentMonth: new Date(),
      selectedDate: new Date()
    };

    this.onDateClick = this.onDateClick.bind(this);
    this.prevMonth = this.prevMonth.bind(this);
    this.nextMonth = this.nextMonth.bind(this);
  }

  onDateClick(day) {
    const date = new Date(day);

    this.setState({
      selectedDate: date.toISOString()
    });
  }

  prevMonth() {
    this.setState({
      currentMonth: subMonths(this.state.currentMonth, 1)
    });
  }

  nextMonth() {
    this.setState({
      currentMonth: addMonths(this.state.currentMonth, 1)
    });
  }

  renderHeader() {
    const dateFormat = 'MMMM yyyy';

    return (
      <div className="header row flex-middle">
        <div className="col col-start">
          <div className="icon" onClick={this.prevMonth}>
            <svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24">
              <path d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z" />
              <path d="M0 0h24v24H0z" fill="none" />
            </svg>
          </div>
        </div>
        <div className="col col-center">
          <span>{format(this.state.currentMonth, dateFormat)}</span>
        </div>
        <div className="col col-end" onClick={this.nextMonth}>
          <div className="icon">
            <svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24">
              <path d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z" />
              <path d="M0 0h24v24H0z" fill="none" />
            </svg>
          </div>
        </div>
      </div>
    );
  }

  renderDays() {
    const dateFormat = 'dddd';
    const days = [];

    const startDate = startOfWeek(this.state.currentMonth);

    for (let i = 0; i < 7; i++) {
      days.push(
        <div className="col col-center" key={i}>
          {format(addDays(startDate, i), dateFormat)}
        </div>
      );
    }

    return <div className="days row">{days}</div>;
  }

  renderCells() {
    const { currentMonth, selectedDate } = this.state;
    const monthStart = startOfMonth(currentMonth);
    const monthEnd = endOfMonth(monthStart);
    const startDate = startOfWeek(monthStart);
    const endDate = endOfWeek(monthEnd);

    const dateFormat = 'd';
    const rows = [];

    let days = [];
    let day = startDate;
    let formattedDate = '';

    while (day <= endDate) {
      for (let i = 0; i < 7; i++) {
        formattedDate = format(day, dateFormat);
        const cloneDay = day;
        days.push(
          <button
            type="button"
            className={`col cell ${
              !isSameMonth(day, monthStart)
                ? 'disabled'
                : isSameDay(day, selectedDate)
                ? 'selected'
                : ''
            }`}
            key={day}
            onClick={() => this.onDateClick(cloneDay)}
            onKeyDown={this.handleKeyDown}
          >
            <span className="number">{formattedDate}</span>
            <span className="bg">{formattedDate}</span>
          </button>
        );
        day = addDays(day, 1);
      }
      rows.push(
        <div className="row" key={day}>
          {days}
        </div>
      );
      days = [];
    }
    return <div className="body">{rows}</div>;
  }

  render() {
    return (
      <div className="calendar">
        {this.renderHeader()}
        {this.renderDays()}
        {this.renderCells()}
      </div>
    );
  }
}

export default Calendar;
