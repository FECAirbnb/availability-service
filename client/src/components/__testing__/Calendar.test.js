/* eslint-disable react/jsx-filename-extension */
/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */
import Enzyme, { shallow, render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import ReactDOM from 'react-dom';
import Calendar from '../Calendar';

Enzyme.configure({ adapter: new Adapter() });

describe('Calendar component', () => {
  it('renders correctly enzyme', () => {
    const wrapper = shallow(<Calendar />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('has the correct state properties', () => {
    const wrapper = shallow(<Calendar />);
    setTimeout(() => {
      expect(wrapper.state('selectedDate')).toEqual(new Date());
      expect(wrapper.state('currentMonth')).toEqual(new Date());
      done();
    }, 1000);
  });

  it('should switch between months', () => {
    const wrapper = shallow(<Calendar />);
    setTimeout(() => {
      wrapper.find('button').simulate('click');
      expect(wrapper.state('currentMonth')).toEqual(subMonths(this.state.currentMonth, 1));
      done();
    }, 1000);
  });
});
