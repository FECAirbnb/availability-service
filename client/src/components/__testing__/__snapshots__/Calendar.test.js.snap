// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar component renders correctly enzyme 1`] = `
<div
  className="calendar"
>
  <div
    className="header row flex-middle"
  >
    <div
      className="col col-start"
    >
      <div
        className="icon"
        onClick={[Function]}
      >
        <svg
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
    </div>
    <div
      className="col col-center"
    >
      <span>
        March 2020
      </span>
    </div>
    <div
      className="col col-end"
      onClick={[Function]}
    >
      <div
        className="icon"
      >
        <svg
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="days row"
  >
    <div
      className="col col-center"
      key="0"
    >
      0022
    </div>
    <div
      className="col col-center"
      key="1"
    >
      0023
    </div>
    <div
      className="col col-center"
      key="2"
    >
      0024
    </div>
    <div
      className="col col-center"
      key="3"
    >
      0025
    </div>
    <div
      className="col col-center"
      key="4"
    >
      0026
    </div>
    <div
      className="col col-center"
      key="5"
    >
      0027
    </div>
    <div
      className="col col-center"
      key="6"
    >
      0028
    </div>
  </div>
  <div
    className="body"
  >
    <div
      className="row"
      key="Sun Mar 08 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
    >
      <button
        className="col cell "
        key="Sun Mar 01 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          1
        </span>
        <span
          className="bg"
        >
          1
        </span>
      </button>
      <button
        className="col cell "
        key="Mon Mar 02 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          2
        </span>
        <span
          className="bg"
        >
          2
        </span>
      </button>
      <button
        className="col cell "
        key="Tue Mar 03 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          3
        </span>
        <span
          className="bg"
        >
          3
        </span>
      </button>
      <button
        className="col cell "
        key="Wed Mar 04 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          4
        </span>
        <span
          className="bg"
        >
          4
        </span>
      </button>
      <button
        className="col cell "
        key="Thu Mar 05 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          5
        </span>
        <span
          className="bg"
        >
          5
        </span>
      </button>
      <button
        className="col cell "
        key="Fri Mar 06 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          6
        </span>
        <span
          className="bg"
        >
          6
        </span>
      </button>
      <button
        className="col cell "
        key="Sat Mar 07 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          7
        </span>
        <span
          className="bg"
        >
          7
        </span>
      </button>
    </div>
    <div
      className="row"
      key="Sun Mar 15 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
    >
      <button
        className="col cell "
        key="Sun Mar 08 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          8
        </span>
        <span
          className="bg"
        >
          8
        </span>
      </button>
      <button
        className="col cell "
        key="Mon Mar 09 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          9
        </span>
        <span
          className="bg"
        >
          9
        </span>
      </button>
      <button
        className="col cell "
        key="Tue Mar 10 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          10
        </span>
        <span
          className="bg"
        >
          10
        </span>
      </button>
      <button
        className="col cell "
        key="Wed Mar 11 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          11
        </span>
        <span
          className="bg"
        >
          11
        </span>
      </button>
      <button
        className="col cell "
        key="Thu Mar 12 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          12
        </span>
        <span
          className="bg"
        >
          12
        </span>
      </button>
      <button
        className="col cell "
        key="Fri Mar 13 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          13
        </span>
        <span
          className="bg"
        >
          13
        </span>
      </button>
      <button
        className="col cell "
        key="Sat Mar 14 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          14
        </span>
        <span
          className="bg"
        >
          14
        </span>
      </button>
    </div>
    <div
      className="row"
      key="Sun Mar 22 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
    >
      <button
        className="col cell "
        key="Sun Mar 15 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          15
        </span>
        <span
          className="bg"
        >
          15
        </span>
      </button>
      <button
        className="col cell "
        key="Mon Mar 16 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          16
        </span>
        <span
          className="bg"
        >
          16
        </span>
      </button>
      <button
        className="col cell "
        key="Tue Mar 17 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          17
        </span>
        <span
          className="bg"
        >
          17
        </span>
      </button>
      <button
        className="col cell "
        key="Wed Mar 18 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          18
        </span>
        <span
          className="bg"
        >
          18
        </span>
      </button>
      <button
        className="col cell "
        key="Thu Mar 19 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          19
        </span>
        <span
          className="bg"
        >
          19
        </span>
      </button>
      <button
        className="col cell "
        key="Fri Mar 20 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          20
        </span>
        <span
          className="bg"
        >
          20
        </span>
      </button>
      <button
        className="col cell "
        key="Sat Mar 21 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          21
        </span>
        <span
          className="bg"
        >
          21
        </span>
      </button>
    </div>
    <div
      className="row"
      key="Sun Mar 29 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
    >
      <button
        className="col cell "
        key="Sun Mar 22 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          22
        </span>
        <span
          className="bg"
        >
          22
        </span>
      </button>
      <button
        className="col cell selected"
        key="Mon Mar 23 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          23
        </span>
        <span
          className="bg"
        >
          23
        </span>
      </button>
      <button
        className="col cell "
        key="Tue Mar 24 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          24
        </span>
        <span
          className="bg"
        >
          24
        </span>
      </button>
      <button
        className="col cell "
        key="Wed Mar 25 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          25
        </span>
        <span
          className="bg"
        >
          25
        </span>
      </button>
      <button
        className="col cell "
        key="Thu Mar 26 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          26
        </span>
        <span
          className="bg"
        >
          26
        </span>
      </button>
      <button
        className="col cell "
        key="Fri Mar 27 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          27
        </span>
        <span
          className="bg"
        >
          27
        </span>
      </button>
      <button
        className="col cell "
        key="Sat Mar 28 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          28
        </span>
        <span
          className="bg"
        >
          28
        </span>
      </button>
    </div>
    <div
      className="row"
      key="Sun Apr 05 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
    >
      <button
        className="col cell "
        key="Sun Mar 29 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          29
        </span>
        <span
          className="bg"
        >
          29
        </span>
      </button>
      <button
        className="col cell "
        key="Mon Mar 30 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          30
        </span>
        <span
          className="bg"
        >
          30
        </span>
      </button>
      <button
        className="col cell "
        key="Tue Mar 31 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          31
        </span>
        <span
          className="bg"
        >
          31
        </span>
      </button>
      <button
        className="col cell disabled"
        key="Wed Apr 01 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          1
        </span>
        <span
          className="bg"
        >
          1
        </span>
      </button>
      <button
        className="col cell disabled"
        key="Thu Apr 02 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          2
        </span>
        <span
          className="bg"
        >
          2
        </span>
      </button>
      <button
        className="col cell disabled"
        key="Fri Apr 03 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          3
        </span>
        <span
          className="bg"
        >
          3
        </span>
      </button>
      <button
        className="col cell disabled"
        key="Sat Apr 04 2020 00:00:00 GMT-0700 (Mountain Standard Time)"
        onClick={[Function]}
        type="button"
      >
        <span
          className="number"
        >
          4
        </span>
        <span
          className="bg"
        >
          4
        </span>
      </button>
    </div>
  </div>
</div>
`;
